File: python/sliding-window/sliding-window-maximum.py
import heapq
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        answer = []
        heap = []
        count = defaultdict(int)
        for i in range(k):
            if nums[i] not in count:
                heapq.heappush(heap, -nums[i])
            count[nums[i]] += 1

        for i in range(k, len(nums)):
            count[nums[i]] += 1
            current = heapq.heappop(heap)
            answer.append(-current)
            count[nums[i - k]] -= 1

            if nums[i - k] == -current:
                while len(heap) > 0:
                    if count[-current] > 0:
                        break
                    current = heapq.heappop(heap)
            if count[-current] > 0:
                heapq.heappush(heap, current)
            heapq.heappush(heap, -nums[i])

        answer.append(-heapq.heappop(heap))

        return answer

File: c++/sliding-window/sliding-window-maximum.cc
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        std::priority_queue<std::pair<int, int>> pq;
        std::vector<int> answer;

        int left = 0;
        for (int i = 0; i < nums.size(); ++i) {
            pq.push(std::make_pair(nums[i], i));

            if (i < k - 1) {
                continue;
            }

            int maximum = std::numeric_limits<int>::min();
            while (maximum == std::numeric_limits<int>::min()) {
                std::pair<int, int> candidate = pq.top();
                if (candidate.second < left) {
                    pq.pop();
                } else {
                    maximum = candidate.first;
                }
            }
            left += 1;

            answer.push_back(maximum);
        }

        return answer;
    }
};
